extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java

```diff
@@ -18,24 +20,50 @@
 public class IcebergTableParquetLocationKey extends ParquetTableLocationKey implements IcebergTableLocationKey {
     private static final String IMPLEMENTATION_NAME = IcebergTableParquetLocationKey.class.getSimpleName();
 
+    /**
+     * Following are derived from the {@link DataFile} that backs the keyed location and are used for ordering of data
+     */
+    private final long dataSequenceNumber;
+    private final long fileSequenceNumber;
+    private final long dataFilePos;
+
+    /**
+     * Following is derived from the {@link ManifestFile} from which the data file was discovered and is used for
+     * ordering of data.
+     */
+    private final long manifestSequenceNumber;
+
     private final ParquetInstructions readInstructions;
 
+    private int cachedHashCode;
+
     /**
      * Construct a new IcebergTableParquetLocationKey for the supplied {@code fileUri} and {@code partitions}.
      *
-     * @param fileUri The file that backs the keyed location
+     * @param dataFile The data file that backs the keyed location
+     * @param manifestFile The manifest file from which the data file was discovered
+     * @param fileUri The {@link URI} for the file that backs the keyed location
      * @param order Explicit ordering index, taking precedence over other fields
      * @param partitions The table partitions enclosing the table location keyed by {@code this}. Note that if this
      *        parameter is {@code null}, the location will be a member of no partitions. An ordered copy of the map will
      *        be made, so the calling code is free to mutate the map after this call
      * @param readInstructions the instructions for customizations while reading
      */
     public IcebergTableParquetLocationKey(
+            @NotNull final DataFile dataFile,
+            @NotNull final ManifestFile manifestFile,
             @NotNull final URI fileUri,
             final int order,
             @Nullable final Map<String, Comparable<?>> partitions,
             @NotNull final ParquetInstructions readInstructions) {
         super(fileUri, order, partitions, readInstructions);
+
+        // Files with unknown sequence numbers should be ordered last.
+        dataSequenceNumber = dataFile.dataSequenceNumber() != null ? dataFile.dataSequenceNumber() : Long.MAX_VALUE;
+        fileSequenceNumber = dataFile.fileSequenceNumber() != null ? dataFile.fileSequenceNumber() : Long.MAX_VALUE;
+        dataFilePos = dataFile.pos() != null ? dataFile.pos() : Long.MAX_VALUE;
+        manifestSequenceNumber = manifestFile.sequenceNumber();
```

Comment:
I don't understand why this is here. If it's for the case where `dataSequenceNumber` and `fileSequenceNumber` are both `null`, my reading suggests that we're not going to get any additional correctness from looking at `manifestSequenceNumber`.

Reply:
So `dataSequenceNumber` and `fileSequenceNumber` can be `null` for old iceberg versions in some cases.
But `manifestSequenceNumber` will never be null for such files too because we always read from some non-null manifest file.
So `manifestSequenceNumber + dataFilePos` will give us some way to order such files.